agent/engine/execcmdagent/manager_linux.go

```diff
@@ -0,0 +1,165 @@
+// +build linux
+
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+package execcmdagent
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"strconv"
+
+	"github.com/cihub/seelog"
+	"github.com/docker/docker/api/types"
+
+	apicontainer "github.com/aws/amazon-ecs-agent/agent/api/container"
+	apitask "github.com/aws/amazon-ecs-agent/agent/api/task"
+	"github.com/aws/amazon-ecs-agent/agent/dockerclient"
+	"github.com/aws/amazon-ecs-agent/agent/dockerclient/dockerapi"
+	"github.com/aws/amazon-ecs-agent/agent/utils/retry"
+)
+
+// RestartIfStopped restarts the ExecCommandAgent in the container passed as parameter, only for ExecCommandAgent-enabled tasks.
+// The status of the ExecCommandAgent in the container is retrieved using a docker exec inspect call, using the dockerExecID
+// stored in container.GetExecCommandAgentMetadata().DockerExecID.
+//
+// If container.GetExecCommandAgentMetadata() is nil, or the ExecCommandAgent is still running, no action is taken.
+// To actually restart the ExecCommandAgent, this function invokes this instance's Start method.
+func (m *manager) RestartIfStopped(ctx context.Context, client dockerapi.DockerClient, task *apitask.Task, container *apicontainer.Container, containerId string) (RestartStatus, error) {
+	if !task.IsExecCommandAgentEnabled() {
+		return NotRestarted, nil
+	}
+	execMD := container.GetExecCommandAgentMetadata()
+	if execMD == nil { // This means m.Start has never been invoked
+		return NotRestarted, nil
+	}
+	res, err := m.inspectExecAgentProcess(ctx, client, execMD)
+	if err != nil {
+		// We don't want to report InspectContainerExec errors, just that we don't know what the status of the agent is
+		return Unknown, nil
+	}
+	if res.Running { // agent still running, nothing to do
+		return NotRestarted, nil
+	}
+	// Restart if not running
+	//TODO: [ecs-exec] retry only for certain exit codes?
```

Comment:
I think we should use the SSM agent systemd service file as our guide:
`Restart=on-failure`
https://github.com/aws/amazon-ssm-agent/blob/master/packaging/linux/amazon-ssm-agent.service#L10
```
If set to on-failure, the service will be restarted when the process exits with a non-zero exit code, is terminated by a signal (including on core dump, but excluding the aforementioned four signals), when an operation (such as service reload) times out, and when the configured watchdog timeout is triggered.
```

Reply:
So, if the ssm-agent exits with a zero value, we don't want to restart it?

Reply:
P.S. I don't think we have access to the signals ssm agent receives through the docker API, all we have is the exit code.

Reply:
Yeah I wonder if we should just always restart n-times, regardless of exit code.  We'll need a good way of surfacing what exit code we're getting.